### PR TITLE
Adjust records behavior around IOperations and analyzer actions.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@mavasani, @cston, @jcouv, @RikkiGibson, @333fred Please review.

**Comment by AlekseyTs** [comment]
@mavasani, @cston, @jcouv, @RikkiGibson, @333fred Please review.

**Comment by AlekseyTs** [comment]
@mavasani, @cston, @jcouv, @RikkiGibson, @333fred Please review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 1]
```diff
@@ -18629,5 +18861,1539 @@ public static void Main()
...
+               new DiagnosticDescriptor("XY0000", "Test", "Test", "Test", DiagnosticSeverity.Warning, true, "Test", "Test");
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(Descriptor);
+
+            public override void Initialize(AnalysisContext context)
+            {
+                context.RegisterSymbolStartAction(Handle1, SymbolKind.Method);
+                context.RegisterSymbolStartAction(Handle1, SymbolKind.Property);
+                context.RegisterSymbolStartAction(Handle1, SymbolKind.Parameter);
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:19379]: Can we add a `context.RegisterSymbolStartAction(..., SymbolKind.NamedType)` with a nested `RegisterSymbolEndAction` and then verify that all the `FireCountX` are `0` at the start of the former callback and `1` at the start of latter callback? This should verify correct ordering of callbacks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 2]
```diff
@@ -18629,5 +18861,1539 @@ public static void Main()
...
+
+            private static readonly DiagnosticDescriptor Descriptor =
+               new DiagnosticDescriptor("XY0000", "Test", "Test", "Test", DiagnosticSeverity.Warning, true, "Test", "Test");
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(Descriptor);
+
+            public override void Initialize(AnalysisContext context)
+            {
+                context.RegisterSymbolAction(Handle, SymbolKind.Method);
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:19299]: Should we also add `context.RegisterSymbolAction(Handle, SymbolKind.NamedType);` to verify callbacks for the record type itself?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 3]
```diff
@@ -2776,6 +2776,9 @@ private static bool IsEquivalentSymbol(ISymbol declaredSymbol, ISymbol otherSymb
                                 Debug.Assert(operationBlock.Parent.IsImplicit);
                                 Debug.Assert(operationBlock.Parent.Parent is IConstructorBodyOperation ctorBody &&
                                     ctorBody.Initializer == operationBlock.Parent);
+
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:2779]: Should we add the below assert?
```suggestion
                                Debug.Assert(!operationBlock.Parent.Parent.IsImplicit);
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 4]
```diff
@@ -5509,7 +5513,6 @@ static void Main()
     }
 }";
             comp = CreateCompilation(sourceB, references: new[] { refA }, parseOptions: TestOptions.RegularPreview, options: TestOptions.ReleaseExe);
-            comp.VerifyDiagnostics();
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:5512]: Was there a reason to remove this call?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs [hunk 5]
```diff
@@ -943,7 +943,20 @@ public void MarkSymbolEndAnalysisComplete(ISymbol symbol, DiagnosticAnalyzer ana
                 {
                     ExecuteBlockActionsCore<CodeBlockStartAnalyzerAction<TLanguageKindEnum>, CodeBlockAnalyzerAction, SyntaxNodeAnalyzerAction<TLanguageKindEnum>, SyntaxNodeAnalyzerStateData, SyntaxNode, TLanguageKindEnum>(
                         codeBlockStartActions, codeBlockActions, codeBlockEndActions, analyzer,
-                        declaredNode, declaredSymbol, executableCodeBlocks, (codeBlocks) => codeBlocks.SelectMany(cb => cb.DescendantNodesAndSelf()),
+                        declaredNode, declaredSymbol, executableCodeBlocks, (codeBlocks) => codeBlocks.SelectMany(
+                            cb =>
+                            {
+                                var filter = semanticModel.GetSyntaxNodesToAnalyzeFilter(cb, declaredSymbol);
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:949]: Is the below existing invocation to this method redundant now? https://github.com/dotnet/roslyn/blob/507c81a5146ba714f827b730cac2339605e22c12/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs#L2695-L2697

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:19299]: > Should we also add context.RegisterSymbolAction(Handle, SymbolKind.NamedType); to verify callbacks for the record type itself?

I can certainly add one, I don't think there is anything interesting about the type itself, the PR is was mostly focused on record members.

---
In reply to: [461193457](https://github.com/dotnet/roslyn/pull/46308#discussion_r461193457) [](ancestors = 461193457)

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs:949]: > Is the below existing invocation to this method redundant now? [https://github.com/dotnet/roslyn/blob/507c81a5146ba714f827b730cac2339605e22c12/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs#L2695-L2697](https://github.com/dotnet/roslyn/blob/507c81a5146ba714f827b730cac2339605e22c12/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs#L2695-L2697)

No, that place iterates through nodes itself

---
In reply to: [461205407](https://github.com/dotnet/roslyn/pull/46308#discussion_r461205407) [](ancestors = 461205407)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:5512]: > Was there a reason to remove this call?

It was determined redundant in one of the previous PRs due to ```.VerifyDiagnostics()``` below.

---
In reply to: [461201750](https://github.com/dotnet/roslyn/pull/46308#discussion_r461201750) [](ancestors = 461201750)

**Comment by AlekseyTs** [comment]
> Only relevant open question is if we can remove the existing invocation to GetSyntaxNodesToAnalyzeFilter in the analyzer driver as we are now invoking it directly within the core analyzer executor.

Responded on that thread.


---
In reply to: [456180326](https://github.com/dotnet/roslyn/pull/46308#issuecomment-456180326) [](ancestors = 456180326)

**Comment by AlekseyTs** [comment]
@cston, @RikkiGibson, @333fred Please review, need a second sign-off.

**Comment by AlekseyTs** [comment]
@cston, @RikkiGibson, @333fred Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 6]
```diff
@@ -178,7 +178,9 @@ internal override IOperation GetOperationWorker(CSharpSyntaxNode node, Cancellat
                 case AccessorDeclarationSyntax accessor:
                     model = (accessor.Body != null || accessor.ExpressionBody != null) ? GetOrAddModel(node) : null;
                     break;
-
+                case RecordDeclarationSyntax { ParameterList: { }, PrimaryConstructorBaseType: { } } recordDeclaration when TryGetSynthesizedRecordConstructor(recordDeclaration) is SynthesizedRecordConstructor ctor:
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:181]: >ctor [](start = 210, length = 4)

Consider omitting the variable name so it's clear we're not expecting to use the value.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:181]: > Consider omitting the variable name so it's clear we're not expecting to use the value.

Sometimes it is useful during debugging

---
In reply to: [461794242](https://github.com/dotnet/roslyn/pull/46308#discussion_r461794242) [](ancestors = 461794242)

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 7]
```diff
@@ -2334,5 +2355,85 @@ internal override Symbol RemapSymbolIfNecessaryCore(Symbol symbol)
...
+                                       else if (node.Parent is BaseListSyntax baseList)
+                                       {
+                                           return node == recordDeclaration.PrimaryConstructorBaseType;
+                                       }
+                                       else if (node.Parent is PrimaryConstructorBaseTypeSyntax baseType && baseType == recordDeclaration.PrimaryConstructorBaseType)
+                                       {
+                                           return node == baseType.ArgumentList;
+                                       }
+
+                                       return true;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:2407]: >return true [](start = 39, length = 11)

Is this correct given the comment above ("Accept only nodes ...")?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 8]
```diff
@@ -14836,6 +14975,13 @@ public Base(int X, int Y)
             Assert.Same("<global namespace>", model.GetEnclosingSymbol(x.SpanStart).ToTestDisplayString());
             Assert.Empty(model.LookupSymbols(x.SpanStart, name: "X"));
             Assert.DoesNotContain("X", model.LookupNames(x.SpanStart));
+
+            var recordDeclarations = tree.GetRoot().DescendantNodes().OfType<RecordDeclarationSyntax>().Skip(1).ToArray();
+
+            Assert.Equal("C", recordDeclarations[0].Identifier.ValueText);
+            Assert.Null(model.GetOperation(recordDeclarations[0]));
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:14982]: >model.GetOperation(recordDeclarations[0]) [](start = 24, length = 41)

Why does `GetOperation()` return `null`?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 9]
```diff
@@ -18629,5 +18861,1601 @@ public static void Main()
...
+                context.RegisterSyntaxNodeAction(Handle4, SyntaxKind.ConstructorDeclaration);
+                context.RegisterSyntaxNodeAction(Handle5, SyntaxKind.PrimaryConstructorBaseType);
+                context.RegisterSyntaxNodeAction(Handle6, SyntaxKind.RecordDeclaration);
+                context.RegisterSyntaxNodeAction(Handle7, SyntaxKind.IdentifierName);
+                context.RegisterSyntaxNodeAction(Handle8, SyntaxKind.SimpleBaseType);
+                context.RegisterSyntaxNodeAction(Handle9, SyntaxKind.ParameterList);
+                context.RegisterSyntaxNodeAction(Handle10, SyntaxKind.ArgumentList);
+            }
+
+            protected void Handle1(SyntaxNodeAnalysisContext context)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:18983]: >protected [](start = 12, length = 9)

Why `protected` for these methods?

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs (hunk 7 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:2407]: > Is this correct given the comment above ("Accept only nodes ...")?

I think yes, this filter is called in the process of descending down the tree. If it returns false for parent node, it is never called for a child node.

---
In reply to: [461820251](https://github.com/dotnet/roslyn/pull/46308#discussion_r461820251) [](ancestors = 461820251)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 8 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:14982]: > Why does GetOperation() return null?

There is no user code for this primary constructor, no code blocks to fire analyzers on, we never create a member semantic model for a primary constructor like that, etc. Technically, we could produce an "empty" IConstructorBodyOperation node, but then we would have to do something with analyzers to make sure an action for this node would fire. Unfortunately, analyzer driver is able to do that only when there is a "real" node for an initializer or a body (see ```GetOperationsToAnalyze``` in AnalyzerDriver.cs), we have neither of them here. I decided to not deal with this complexity right now, when we don't really have a real scenario to enable.  

---
In reply to: [461876729](https://github.com/dotnet/roslyn/pull/46308#discussion_r461876729) [](ancestors = 461876729)

**Pull request merged**